modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java

```diff
@@ -720,26 +705,68 @@ else if (customMsg instanceof WalStateAbstractMessage
         }
     }
 
-    /**
-     * @param msg Event message.
-     * @param evtType Event type.
-     */
-    private void recordEvent(String msg, int evtType) {
-        GridKernalContext ctx = cctx.kernalContext();
+    /** */
+    private void onClusterStateChangeFinish(
+        IgniteInternalFuture<AffinityTopologyVersion> fut,
+        ExchangeActions exchActions,
+        boolean baselineChanging
+    ) {
+        A.notNull(exchActions, "exchActions");
 
-        Collection<ClusterNode> srvNodes = ctx.cluster().get().forServers().nodes();
+        GridEventStorageManager evtMngr = cctx.kernalContext().event();
 
-        Collection<BaselineNode> baselineNodes = new ArrayList<>(srvNodes);
+        if (exchActions.activate() && evtMngr.isRecordable(EVT_CLUSTER_ACTIVATED) ||
+            exchActions.deactivate() && evtMngr.isRecordable(EVT_CLUSTER_DEACTIVATED) ||
+            exchActions.changedClusterState() && evtMngr.isRecordable(EVT_CLUSTER_STATE_CHANGED) ||
+            baselineChanging && evtMngr.isRecordable(EVT_BASELINE_CHANGED)
+        ) {
+            List<Event> evts = new ArrayList<>(2);
 
-        ctx.getStripedExecutorService().execute(CLUSTER_ACTIVATION_EVT_STRIPE_ID, new Runnable() {
-            @Override public void run() {
-                if (ctx.event().isRecordable(evtType)) {
-                    ClusterActivationEvent evt = new ClusterActivationEvent(ctx.discovery().localNode(), msg, evtType, baselineNodes);
+            ClusterNode locNode = cctx.kernalContext().discovery().localNode();
 
-                    ctx.event().record(evt);
-                }
+            Collection<BaselineNode> bltNodes = cctx.kernalContext().cluster().get().currentBaselineTopology();
+
+            boolean collectionUsed = false;
+
+            if (exchActions.activate() && evtMngr.isRecordable(EVT_CLUSTER_ACTIVATED)) {
+                assert !exchActions.deactivate() : exchActions;
+
+                collectionUsed = true;
+
+                evts.add(new ClusterActivationEvent(locNode, "Cluster activated.", EVT_CLUSTER_ACTIVATED, bltNodes));
```

Comment:
can we use here : ClusterStateChangeEvent ?

Reply:
We must support backward compatibility.